### PR TITLE
810 download experiment

### PR DESCRIPTION
**Pull request by davidsmejia**
## Issue Number

#810 

## Purpose/Implementation Notes

Adds some helpers for moving away from redux
Adds a DownloadExperiment component
Adds a composable DownloadOptionsForm component

## Types of changes

What types of changes does your code introduce?

* [ ] Bugfix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Functional tests

Downloaded experiment as a dataset, refreshed to see if the badge was correctly showing

## Checklist

* [x] Lint and unit tests pass locally with my changes
* [ ] I have added tests that prove my fix is effective or that my feature works
* [ ] I have added necessary documentation (if appropriate)
* [x] Any dependent changes have been merged and published in downstream modules

## Screenshots

Please attach any screenshots that illustrate these changes.

### CODE REVIEW

File: src/common/helpers.js [hunk 1]
```diff
@@ -359,3 +359,23 @@ export function numberFormatter(num, digits = 0) {
...
   return (num / si[i].value).toFixed(digits).replace(rx, '$1') + si[i].symbol;
 }
+
+/**
+ * Subscribe an email to CCDL Emails
+ *
+ * @param {*} EmailAddress
+ */
+
+export const subscribeToMailingList = async email => {
```
Comment (arielsvn) [inline src/common/helpers.js]: Do we use this anywhere?

File: src/common/helpers.js [hunk 2]
```diff
@@ -359,3 +359,23 @@ export function numberFormatter(num, digits = 0) {
...
+  const portalId = '5187852';
+  const formGuid = 'ea5b970d-c22f-4b42-b569-26f84b1cea2f';
+  const formUrl = `https://api.hsforms.com/submissions/v3/integration/submit/${portalId}/${formGuid}`;
+  try {
+    await Ajax.post(formUrl, {
+      fields: [{ name: 'email', value: email }],
+    });
+  } catch (err) {
+    // Show error message?"
```
Comment (arielsvn) [inline src/common/helpers.js]: We should log the failure to somewhere, maybe sentry?

File: src/common/hooks.js [hunk 3]
```diff
@@ -78,7 +78,11 @@ export function useLocalStorage(key, initialValue) {
       setStoredValue(valueToStore);
       // Save to local storage
       if (!isServer) {
-        window.localStorage.setItem(key, JSON.stringify(valueToStore));
+        if (valueToStore === null) {
```
Comment (arielsvn) [inline src/common/hooks.js]: Should this be `valueToStore === undefined`? Why not save `null` values?

File: src/components/DatasetDownloadOptionsForm.js [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+
+// This is the mechanism for propagating properties to child components
+// TODO: Any option should be overridable at the child component level
+const DatasetDownloadOptionsContext = React.createContext();
+
+// If you do not add children this is what will be populated
+// Notice no EMAIL
+const DefaultOptions = () => {
+  return (
+    <DatasetDownloadOptionsContext.Consumer>
```
Comment (arielsvn) [inline src/components/DatasetDownloadOptionsForm.js]: Is it better to use the hook version of this [useContext](https://reactjs.org/docs/hooks-reference.html#usecontext)?

File: src/components/DownloadExperiment.js [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+const DownloadExperiment = ({ experiment, showProcessing = false }) => {
+  const [emailAddress] = useLocalStorage('email-address');
+
+  const {
+    accession_code: experimentAccessionCode,
+    num_downloadable_samples: downloadableSamplesCount,
+  } = experiment;
+
+  const data = {};
+  data[experimentAccessionCode] = experiment.samples.map(s => s.accession_code);
```
Comment (arielsvn) [inline src/components/DownloadExperiment.js]: It's also possible to initialize the property directly:

```suggestion
  const data = {
    [experimentAccessionCode]: experiment.samples.map(s => s.accession_code)
  };
```

File: src/components/DownloadExperiment.js [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+    };
+
+    const justStarted = dataset.id !== processingDataset;
+    const wasProccessing = processingDataset && !dataset.id;
+
+    if (datasetId && (justStarted || wasProccessing)) refreshState();
+  }, [dataset, setDataset, processingDataset, setDatasetIsProcessing]);
+
+  // Hide this component
+  if (downloadableSamplesCount === 0) return false;
```
Comment (arielsvn) [inline src/components/DownloadExperiment.js]: Should this be the responsability of the parent? To only show the component when there're downloadable samples?

File: src/components/DownloadExperiment.js [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+      if (refreshedDataset.is_processing) {
+        setDataset(refreshedDataset);
+        setDatasetIsProcessing(refreshedDataset.id);
+      } else {
+        setDataset(emptyDataset);
+        setDatasetIsProcessing(null);
+      }
+    };
+
+    const justStarted = dataset.id !== processingDataset;
```
Comment (arielsvn) [inline src/components/DownloadExperiment.js]: in what scenario is this `false`?

**Comment by arielsvn** [comment]
Also, when the "Download Now" button next to the samples table is used to download, the experiment is not marked as Processing.

File: src/common/helpers.js (hunk 1 above)
Comment (davidsmejia) [reply src/common/helpers.js]: good call that should be at the end of the submit for datasetdowloadoptionsform.js

File: src/common/helpers.js (hunk 2 above)
Comment (davidsmejia) [reply src/common/helpers.js]: we talked about this before... this is how it is currently implemented but im open to telling sentry.. is there any problem with passing the email or we could pass the dataset and then manually subscribe them to the mailing list if it fails, but I think it's very unlikely to fail unless maybe they have a weird internet connection drop

Comment (arielsvn) [reply src/common/helpers.js]: yes, sending the dataset id should be enough. Hopefully, this never fails

File: src/common/hooks.js (hunk 3 above)
Comment (davidsmejia) [reply src/common/hooks.js]: Yeah im open to either one... or we could also pass back an array of 3 values with an explicit clear function so something like

```
const [token, setToken, clearToken] = useLocalStorage('token', false)
```

File: src/components/DatasetDownloadOptionsForm.js (hunk 4 above)
Comment (davidsmejia) [reply src/components/DatasetDownloadOptionsForm.js]: I like this better, good look!

File: src/components/DownloadExperiment.js (hunk 6 above)
Comment (davidsmejia) [reply src/components/DownloadExperiment.js]: yeah that would be a good call for this case (actually I already think it's implemented for the case when there are no downloadable samples, but I think we should still handle the case where we pass a parameter to hide when the dataset is being processed and the button shouldn't show, Just to keep the logic more dry...

File: src/components/DownloadExperiment.js (hunk 7 above)
Comment (davidsmejia) [reply src/components/DownloadExperiment.js]: This is false when the id's are the same, so the form just completed and updated both values to be the same.

Comment (davidsmejia) [reply src/components/DownloadExperiment.js]: This is false when the processing dataset is the same as the local storage dataset. It will be false when the page is refreshed.

**Comment by davidsmejia** [comment]
> Also, when the "Download Now" button next to the samples table is used to download, the experiment is not marked as Processing.

Yeah by design:
https://user-images.githubusercontent.com/15315514/74863463-f05bec80-531b-11ea-8e23-3f242108d65f.png

**Comment by dvenprasad** [comment]
Hey! I looked at the preview and noticed some things were a little different from design, and wanted know what the rationale was.

1) Search result cards don't have `download now` buttons (maybe we talked about this? not sure) 

2) Download Now Modal:
    - Shows option to skip QN for microarray experiments. This option is only applicable to RNA-seq samples. (Can we disable it for micoarray experiments?) 
     - Shows aggregate options for experiments with only one species. (Can we disable it for single species experiment?) 

There were some aesthetic inconsistencies I noticed as well for the `Download Now Modal`
- The padding on the modal is too small. It should be 40px. 
- The spacing between label and form field is also too small. It should be 16px.

**Comment by davidsmejia** [comment]
@arielsvn could you take another look at this, I was able to support the requested feature set by pulling some data management outside of the component, would love to get your thoughts on it. going to tag you for review

some things that should be working
* synchronization between navigation - download from search will show tag on experiment page
* updates on the same page - ie download form near samples section should update the top download bar too
* synchronization between tabs - download in one tab should update state on the other tab (if its the same experiment)

File: src/common/helpers.js (hunk 2 above)
Comment (davidsmejia) [reply src/common/helpers.js:378]: I'm not really sure how we would escalate exceptions to Sentry without triggering an error boundary capture, do you have any direction for this? Also let's make it a new ticket because this isn't really a change specific to this PR it's just moving the same logic we had so it can be shared.

**Comment by dvenprasad** [comment]
@arielsvn we email them the dataset. On the `We're processing` page(?)/state(?) of the modal, we provide a link to track progress, which opens up a new tab.

File: src/components/DatasetDownloadOptionsForm.js [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+    >
+      <Button type="submit" text={actionText} disabled={isSubmitting} />
+    </div>
+  );
+};
+
+/**
+ *
+ */
+const DatasetDownloadOptionsFormComponent = ({
```
[the remark this answers is not in the document]
Comment (davidsmejia) [reply src/components/DatasetDownloadOptionsForm.js:251]: Yeah this is setting up for removing redux for managing the original dataset. It's a lot of duplication but I'm not going to be including the refactor for that inside of this because it will require a lot more time to untangle that flow.

File: src/components/DownloadExperiment.js [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+  EmailAddressOptions,
+  SubmitDatasetOptionsForm,
+} from './DatasetDownloadOptionsForm';
+
+import { getDataSet } from '../api/dataSet';
+
+import { useLocalStorage, useWatchedLocalStorage } from '../common/hooks';
+import { getDomain } from '../common/helpers';
+
+// TODO: Rethink how we use email fallback to support designs
```
[the remark this answers is not in the document]
Comment (davidsmejia) [reply src/components/DownloadExperiment.js:20]: We don't have user accounts so we assume that the email in local store. Now we want to support multiple simultaneous datasets which potentially could be associated with multiple email addresses. The problem is that we don't receive the email from the API (for privacy reasons).

File: src/components/DatasetDownloadOptionsForm.js [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+  token,
+  createToken: connectedCreateToken,
+}) => {
+  const [showAdvancedOptions, setAdvancedOptions] = React.useState(
+    !dataset.quantile_normalize || advancedOptions
+  );
+
+  const [emailAddress, setEmailAddress] = useLocalStorage('email-address', '');
+
+  const defaultValidation = () => {
```
[the remark this answers is not in the document]
Comment (davidsmejia) [reply src/components/DatasetDownloadOptionsForm.js]: I think it might be possible but this current implementation requires that the token is passed down into the component which is used in the validation rules. I moved it outside and have a passable parameter now.

File: src/common/hooks.js [hunk 11]
```diff
@@ -89,6 +93,71 @@ export function useLocalStorage(key, initialValue) {
   return [storedValue, setValue];
 }
 
+/**
+ * Same as useLocalStrage but updates state when local storage changes
+ */
+
+export function useWatchedLocalStorage(key, initialValue) {
```
[the remark this answers is not in the document]
Comment (davidsmejia) [reply src/common/hooks.js:100]: We might be able to replace it but that would be outside the scope of this PR I'm not sure what would break if it is expecting the existing the behavior.

File: src/api/dataSet.js [hunk 12]
```diff
@@ -57,6 +64,14 @@ export async function updateDataSet(dataSetId, dataSet, details = false) {
     : result;
 }
 
+export async function updateDataSetOptions(dataset, token = false) {
```
[the remark this answers is not in the document]
Comment (davidsmejia) [reply src/api/dataSet.js:67]: The one above it wasn't taking headers and had a weird return clause. I think it's better to leave this alone for now. This will be removed anyway when we move away from redux.

File: src/pages/experiments/index.js [hunk 13]
```diff
@@ -106,18 +113,29 @@ let Experiment = ({
...
+                        all: true,
+                        total: numDownloadableSamples,
+                      },
+                    }}
+                  />
+                  {!dataset.is_processing ? (
+                    <DownloadExperiment
+                      experiment={experiment}
+                      dataset={dataset}
+                      setDataset={setDataset}
```
[the remark this answers is not in the document]
Comment (davidsmejia) [reply src/pages/experiments/index.js:133]: * The experiment page wants to be able to manage the state (in order to update multiple instances of the same dataset being downloadable) 
* I think its better if we move that management outside of the component, dataset and setDataset can also be wrapped this way
* This configuration is just one, probably one of the more complex examples, but in the future other components won't be so deeply nested (ie) dataset download page with will just need to have the dataset its configuring options for and a way to update it because of the way the API works, ideally we will move away from that in the future and can treat it more restfully. 📅
